### PR TITLE
Fix for issue #556.

### PR DESCRIPTION
**Pull request by ILMTitan**
Fix for issue #556.
Filter all repository-changes.* topics.
For gcr topics, also filter with ':' in project ids replaced by '%2F'.

### CODE REVIEW

File: GoogleCloudExtension/GoogleCloudExtension/CloudExplorerSources/PubSub/PubSubSourceRootViewModel.cs [hunk 1]
```diff
@@ -162,21 +163,24 @@ public override void InvalidateProjectOrAccount()
...
+        /// A regex string to match a blacklisted topic.
         /// </returns>
-        private string FormatBlacklistedTopics(string blacklistedTopicString)
+        private IEnumerable<string> FormatBlacklistedTopics(string blacklistedTopicString)
         {
-            return BlackListPrefix + string.Format(blacklistedTopicString, CurrentProjectId);
+            string restUrlPrefix = Regex.Escape(string.Format(BlackListPrefix, CurrentProjectId));
+            yield return restUrlPrefix + string.Format(blacklistedTopicString, CurrentProjectId);
+            yield return restUrlPrefix +
+                string.Format(blacklistedTopicString, Uri.EscapeDataString(CurrentProjectId.Replace(":", "/")));
```
Comment (Deren-Liao) [inline GoogleCloudExtension/GoogleCloudExtension/CloudExplorerSources/PubSub/PubSubSourceRootViewModel.cs]: From the bug,  only repository-changes needs to use this regex.    I don't think  you need to do IsMatch on every other strings like cloud-builds:projects:xxxx-id-xxx:topics

File: GoogleCloudExtension/GoogleCloudExtension/CloudExplorerSources/PubSub/PubSubSourceRootViewModel.cs [hunk 2]
```diff
@@ -162,21 +163,24 @@ public override void InvalidateProjectOrAccount()
...
+            return !s_blacklistedTopics.SelectMany(FormatBlacklistedTopics).Any(s => Regex.IsMatch(topic.Name, s));
         }
 
         /// <summary>
-        /// Gets the full formatted name of a blacklisted topic given a blacklisted topic format string.
+        /// Gets a regex to filter the blacklisted topics.
         /// </summary>
         /// <param name="blacklistedTopicString">
-        /// A format string of a blacklisted topic. Needs a project id as input.
+        /// A format string of a blacklisted topic. It may take project id as the first format arg.
```
Comment (Deren-Liao) [inline GoogleCloudExtension/GoogleCloudExtension/CloudExplorerSources/PubSub/PubSubSourceRootViewModel.cs]: Nit: Add examples on the final filter string.   It takes some effort to figure out what are the final filter strings look like.

Comment (ILMTitan) [inline GoogleCloudExtension/GoogleCloudExtension/CloudExplorerSources/PubSub/PubSubSourceRootViewModel.cs]: Done.

File: GoogleCloudExtension/GoogleCloudExtension/CloudExplorerSources/PubSub/PubSubSourceRootViewModel.cs (hunk 1 above)
Comment (ILMTitan) [inline GoogleCloudExtension/GoogleCloudExtension/CloudExplorerSources/PubSub/PubSubSourceRootViewModel.cs]: Added anchors to the regexs.

**Comment by ILMTitan** [comment]
PTAL

File: GoogleCloudExtension/GoogleCloudExtension/CloudExplorerSources/PubSub/PubSubSourceRootViewModel.cs [hunk 3]
```diff
@@ -162,21 +163,29 @@ public override void InvalidateProjectOrAccount()
...
-        private string FormatBlacklistedTopics(string blacklistedTopicString)
+        private IEnumerable<string> FormatBlacklistedTopics(string blacklistedTopicString)
         {
-            return BlackListPrefix + string.Format(blacklistedTopicString, CurrentProjectId);
+            string escapedProjectId = Regex.Escape(CurrentProjectId);
+            string restUrlPrefix = string.Format(BlackListPrefix, escapedProjectId);
+            yield return restUrlPrefix + string.Format(blacklistedTopicString, escapedProjectId);
+            yield return restUrlPrefix +
+                string.Format(
+                    blacklistedTopicString, Regex.Escape(Uri.EscapeDataString(CurrentProjectId.Replace(":", "/"))));
```
Comment (Deren-Liao) [inline GoogleCloudExtension/GoogleCloudExtension/CloudExplorerSources/PubSub/PubSubSourceRootViewModel.cs]: Nit: give an example on this too 
Uri.EscapeDataString(CurrentProjectId.Replace(":", "/"))
 
Do you have unit test that covers this too?

**Pull request merged**